### ISSUE TITLE
File is stored both in tmp and destination directory during upload and occupies space twice

### ISSUE THREAD

**Issue by haziegler** [nextcloud/server#7920]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce

I discovered that during large file uploads, NextCloud temporary occupies space twice: Once in the final directory where the user uploads the file, and the same in the temporary directory.

This is a snapshot of both directories while uploading a 1 GB file:
```
$ ls -lai tmp/* data/my_username/files/test* -h
2117875 -rw-r--r--. 1 apache apache 914M 17. Jan 15:55 data/my_username/files/test.ocTransferId1441337350.part
3145982 -rw-------. 1 apache apache 914M 17. Jan 15:55 tmp/php3Dri89
```

As you can see, the file is temporarily written into the destination directory and in tmp. The files have always the same size during the upload, which means that it's written at the same time. Both directories are on the same mountpoint, however the inode is different as you can see. So these are really two files. This means, during the upload, the free space must be minimum double of the file size you upload.

### Steps to reproduce
1. Upload a large file and monitor the tmp and destination directory

### Expected behaviour
The file should either be written into the tmp directory and after the upload moved to the final directory. Alternatively, the upload could be directly into the final directory using a temporary file name and then renamed.

### Actual behaviour
File is written during the upload into tmp and final directory at the same time. After the successful upload, the file in tmp dissapears and the one in the final directory is renamed to its real name.

### Server configuration
**Operating system**: Fedora 27

**Web server:** httpd-2.4.29

**Database:** sqlite-3.20.1

**PHP version:** php-7.1.13

**Nextcloud version:** 12.0.4

**Updated from an older Nextcloud/ownCloud or fresh install:** Installed with OC 9.x, periodically updated, migrated to NC 10 or 11. Later updated to 12.

**Where did you install Nextcloud from:** https://download.nextcloud.com/server/releases/

**Are you using external storage, if yes which one:** No

**Are you using encryption:** No

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** Firefox 57.0.4

**Operating system:** Fedora 27

**Comment by MorrisJobke** [comment]
cc @rullzer @icewind1991 Might be interesting.

**Comment by rullzer** [comment]
mmm I can't reproduce this.

**Comment by haziegler** [comment]
Did you try to reproduce the problem on NextCloud 13? Then I think the reason you could't preproduce it was that one of the folders is now different. However, the problem still exists. Instead of data/my_username/files/ and the tmp folder (in pre-13), the file is now uploaded at the same time to data/myusername/uploads/web-file-upload-.../ and the tmp folder:

```
# ls -la tmp/ data/myusername/uploads/web-file-upload-4539330648b80f94ef3bf911f6d77ac9-1520166932223/
data/myusername/uploads/web-file-upload-4539330648b80f94ef3bf911f6d77ac9-1520166932223/:
insgesamt 231364
drwxr-xr-x. 2 apache apache     4096  4. Mär 13:38 .
drwxr-xr-x. 3 apache apache       76  4. Mär 13:35 ..
-rw-r--r--. 1 apache apache 10485760  4. Mär 13:37 188743680
-rw-r--r--. 1 apache apache 10485760  4. Mär 13:38 220200960
-rw-r--r--. 1 apache apache  8806400  4. Mär 13:38 220200960.ocTransferId805739011.part

tmp/:
insgesamt 16320
drwx------. 2 apache apache      23  4. Mär 13:38 .
drwx------. 4 apache apache     116  9. Feb 16:00 ..
-rw-------. 1 apache apache 8806400  4. Mär 13:38 phpwii5lG
```

Another difference I see in NC13 is, that NC13 uploads the file in 10MB chunks to both folders. In tmp, maximum only one file up to 10MB exists, while in web-file-upload*, all chunk files are stored and merged at the end. First I thought this would solve the need of having uploaded_file_size*2 of free space available, but this isn't the case.

Of course, the 10MB chunk upload make sense to bypass certain webserver limits. However, in the web-file-upload directory, if you store all of the chunk files and merge them at the end, also the amount of available free space must be larger than the double of the uploaded file size, because temporarily, all chunk files and the final file exist at the same time.

My suggestion to solve this problem is to keep the 10MB chunk upload behaviour to tmp. However, each time 10MBs are uploaded, append the content to the _one_ file in web-file-upload and delete the chunk in tmp before the next 10MB chunk upload starts. This way only the uploaded file size + 10 MB free space needs to be at least available.

**Comment by bendangelo** [comment]
Is there an easy work around for this? The uploads directory is taking up a lot of space. Can I just delete all contents after uploading everything.

**Comment by fracarvic** [comment]
It seems that this is the normal behavior of PUT uploads with php, nothing related to nextcloud.

I try with a clean docker container with php:7-apache image and put this file as upload.php

```
<?php
$datosPUT = fopen("php://input", "r");
$fp = fopen("/tmp/uploaded_file", "w");
while ($datos = fread($datosPUT, 1024))
  fwrite($fp, $datos);
fclose($fp);
fclose($datosPUT);
?>
```

and try upload a big file with `curl -T big_file.iso localhost:80/upload.php`, and while uploading appears a file `/tmp/phpXXXXX` growing exactly same as `/tmp/uploaded_file`. When upload has finished `/tmp/phpXXXXX` disappears.

Tested with php:5-apache and same behavior.

I didn't find any php.ini options to prevent file caching of PUT requests, I need 30 GB free to upload 15GB file.

**Comment by sooslaca** [comment]
yep, still happening in version 17.0. Being there in two places, consuming 2X space and 2X io.

```
6c03af6471a3:/var/www/html# ls -la /tmp | grep php
-rw-------    1 www-data www-data 38481747968 Sep 27 14:44 phpbfmDaj

6c03af6471a3:/var/www/html/data/xxxxxxxxx# ls -la | grep <filename>
-rw-r--r--    1 www-data www-data 40681603072 Sep 27 14:46 bigfile.txt
```

Usingdocker image
```
nextcloud                        fpm-alpine          c03acf590307        12 hours ago        352MB
```

**Comment by acsfer** [comment]
Able to reproduce here too. `/tmp` is the folder used by Nextcloud and PHP. Uploading a 20GB file produces **one file** of 20GB and **20 files** of 512Mb each. So to upload 20GB, 40GB of temp disk is needed...otherwise upload fails with messages like `Expecting filesize of`...

## Steps to reproduce

- NC 19.0 beta 4
- PHP 7.3.17

1. `fallocate -l 20G 20G.img` to generate a 20GB file
2. `curl -uv test:test -T 20G.img "http://test.lan/remote.php/dav/files/test/20G.img"` to upload it
(wait for `We are completely uploaded and fine here` returned by curl) and issue
3. `df -h /tmp`

```
root@nc:/tmp# ll
total 44231688
drwxrwxrwt  9 root    root            4096 Apr 21 17:08 ./
drwxr-xr-x 23 root    root            4096 Apr 21 13:05 ../
-rw-------  1 test www-data   524288000 Apr 21 17:07 php0Bl5R0
-rw-------  1 test www-data   524288000 Apr 21 17:04 php1I3GXF
-rw-------  1 test www-data   524288000 Apr 21 17:02 php5hBopR
-rw-------  1 test www-data   524288000 Apr 21 17:07 php6GdBl1
-rw-------  1 test www-data   524288000 Apr 21 17:02 php8SoZP8
-rw-------  1 test www-data   524288000 Apr 21 16:55 php94C1P7
-rw-------  1 test www-data   524288000 Apr 21 17:01 phpAj4E6K
-rw-------  1 test www-data   524288000 Apr 21 17:00 phpbFMioX
-rw-------  1 test www-data   524288000 Apr 21 16:57 phpbpMui8
-rw-------  1 test www-data   524288000 Apr 21 17:08 phpbr44Fm
-rw-------  1 test www-data   524288000 Apr 21 17:00 phpC7eAlS
-rw-------  1 test www-data   524288000 Apr 21 17:06 phpDqkjJA
-rw-------  1 test www-data   503316480 Apr 21 17:08 phpE3WWF8
-rw-------  1 test www-data   524288000 Apr 21 16:59 phpgg58CF
-rw-------  1 test www-data   524288000 Apr 21 16:57 phpgTn7yU
-rw-------  1 test www-data   524288000 Apr 21 16:59 phpH5Rofn
-rw-------  1 test www-data   524288000 Apr 21 17:02 phpHB2eXd
-rw-------  1 test www-data   524288000 Apr 21 16:58 phphB2RRt
-rw-------  1 test www-data   524288000 Apr 21 17:04 phpHFsJ93
-rw-------  1 test www-data   524288000 Apr 21 17:06 phpIXaJuR
-rw-------  1 test www-data   524288000 Apr 21 17:01 phpjM1Odf
-rw-------  1 test www-data 21474836480 Apr 21 17:08 phpkbPZho
-rw-------  1 test www-data   524288000 Apr 21 16:57 phpkpFvIV
-rw-------  1 test www-data   524288000 Apr 21 17:07 phpKxwy4v
-rw-------  1 test www-data   524288000 Apr 21 16:55 phpliytUl
-rw-------  1 test www-data   524288000 Apr 21 17:06 phpLvOcDN
-rw-------  1 test www-data   524288000 Apr 21 17:01 phpLXjWwW
-rw-------  1 test www-data   524288000 Apr 21 16:59 phpMIhuqB
-rw-------  1 test www-data   524288000 Apr 21 16:57 phpORClLm
-rw-------  1 test www-data   524288000 Apr 21 16:55 phpPmZnEN
-rw-------  1 test www-data   524288000 Apr 21 17:02 phprygofc
-rw-------  1 test www-data   524288000 Apr 21 17:05 phps1bH65
-rw-------  1 test www-data   524288000 Apr 21 17:07 phptMDOMB
-rw-------  1 test www-data   524288000 Apr 21 17:00 phpU6EVla
-rw-------  1 test www-data   524288000 Apr 21 16:55 phpubJd7r
-rw-------  1 test www-data   524288000 Apr 21 17:03 phpuDOvG2
-rw-------  1 test www-data   524288000 Apr 21 16:55 phpVoUm9U
-rw-------  1 test www-data   524288000 Apr 21 17:05 phpwL1OFG
-rw-------  1 test www-data   524288000 Apr 21 17:04 phpwN7SBU
-rw-------  1 test www-data   524288000 Apr 21 17:06 phpyi8xnZ
-rw-------  1 test www-data   524288000 Apr 21 17:01 phpZ44p88
-rw-------  1 test www-data   524288000 Apr 21 17:05 phpzoLawe
```

I was expecting `phpkbPZho` file only, not the other ones.

**Comment by cynecx** [comment]
Is this php behavior documented somewhere? Or is there an upstream bug report?

**Comment by mrlaojia** [comment]
the bug still happening in version 20.0.1， it's insufferable  when i upload a large file

**Comment by panki27** [comment]
Running Nextcloud 19.01, facing the same problem. It seems that the temp directory never gets cleared.

**Comment by szaimen** [comment]
Hi, please update to 24.0.8 or better 25.0.2 and report back if it fixes the issue. Thank you!

**Comment by topas-rec** [comment]
Tested a 2 GB file upload on 25.0.0.

tmp folder is no longer present in `/var/www/html`. So I was unable to monitor this folder.

Guessed the upload will be temporarily stored somewhere else. Found a large file with changing name and size in  `/tmp` called `phpXXXXXX`. (replace X with random numbers and letters)
This file was no longer there after the upload has finished. (Also no file in `/var/www/html/tmp` after the upload has finished)

To be sure the `html` folder has not increased after the upload, I uploaded the file a second time and measured the `html` folder size before and after the upload.
There was no increase larger than 100MB measurable of the `html` folder. (Nautlius only showed GG.M one decimal place for hundreds of megabytes and it did not change after the upload)

So I cannot reproduce this issue anymore inside the `html` folder.
